src/coreclr/jit/lclvars.cpp

```diff
@@ -5853,642 +5878,166 @@ void Compiler::lvaUpdateArgsWithInitialReg()
-        // Struct parameter 'a3' is 16-bytes that is required to be double aligned;
-        //  the caller skips [OutArg+4] and starts the argument at [OutArg+8].
-        // Thus the caller generates the correct code to pass the arguments.
-        // When generating code to receive the arguments we set codeGen->regSet.rsMaskPreSpillRegArg to [r1,r2,r3]
-        //  and spill these three registers as the first instruction in the prolog.
-        // Then when we layout the arguments' stack offsets we have an argOffs 0 which
-        //  points at the location that we spilled r1 into the stack.  For this first
-        //  struct we take the lvIsRegArg path above with "codeGen->regSet.rsMaskPreSpillRegArg &" matching.
-        // Next when we calculate the argOffs for the second 16-byte struct we have an argOffs
-        //  of 16, which appears to be aligned properly so we don't skip a stack slot.
-        //
-        // To fix this we must recover the actual OutArg offset by subtracting off the
-        //  sizeof of the PreSpill register args.
-        // Then we align this offset to a multiple of 8 and add back the sizeof
-        //  of the PreSpill register args.
-        //
-        // Dev11 Bug 71767: failure of assert(sizeofPreSpillRegArgs <= argOffs)
-        //
-        // We have a method with 'this' passed in r0, RetBuf arg in r1, VarArgs cookie
-        // in r2. The first user arg is a 144 byte struct with double alignment required,
-        // r3 is skipped, and the struct is passed on the stack. However, 'r3' is added
-        // to the codeGen->regSet.rsMaskPreSpillRegArg mask by the VarArgs cookie code, since we need to
-        // home all the potential varargs arguments in registers, even if we don't have
-        // signature type information for the variadic arguments. However, due to alignment,
-        // we have skipped a register that doesn't have a corresponding symbol. Make up
-        // for that by increasing argOffs here.
-        //
-
-        int sizeofPreSpillRegArgs = genCountBits(codeGen->regSet.rsMaskPreSpillRegs(true)) * REGSIZE_BYTES;
-
-        if (argOffs < sizeofPreSpillRegArgs)
+        else
         {
-            // This can only happen if we skipped the last register spot because current stk arg
-            // is a struct requiring alignment or a pre-spill alignment was required because the
-            // first reg arg needed alignment.
+            // Some ABIs may split parameters across registers and stack:
             //
-            // Example 1: First Stk Argument requiring alignment in vararg case (same as above comment.)
-            //            Signature (int a0, int a1, int a2, struct {long} a3, ...)
+            // - On Windows, the Arm64 varargs ABI can split a 16 byte struct across x7 and stack
+            // - Arm32 generally allows structs to be split
+            // - LA64/RISCV64 both allow splitting of 16-byte structs across 1 register and stack
```

Comment:
For the splitting, is only the register saved on the stack within the prolog?
Should the second part of the struct which passed on caller's stack be copyed to on the prolog stack?
If don't copy the second part, how to process this part as the two parts are not stored on a continue stack space.

Reply:
Yes, for arm32/arm64 we only need to store the register part. The stack segments are always at the beginning and will be contiguous with the register segments. I thought LA64/RV64 was similar here.
@tomeksowi is enabling code in #101288 to handle it more generally. LA64 can do the same.

Out of curiosity, can you show the ABI information JITDUMO for the case with the split segments that doesn't have the stack segment first? For win-arm64 I think the split args were deliberately designed this way to make varargs possible to handle.

Reply:
For the old-style ABI, LA64 will copy the second part to the prolog stack where the splitting struct's size is the whole size and had allocated the whole stack space, that is, we can copy the second part directly.

I'm finding the test case, but I think/doubt the new-style can not process this case as we expected liking the old-style.

Reply:
```
;  V00 arg0         [V00    ] (  1,  1   )     int  ->  [fp+0x54]  do-not-enreg[]
;  V01 arg1         [V01    ] (  1,  1   )  struct (16) [fp+0x40]  do-not-enreg[XS] addr-exposed ld-addr-op <Point>
;  V02 arg2         [V02    ] (  1,  1   )  struct (16) [fp+0x30]  do-not-enreg[XS] addr-exposed ld-addr-op <Point>
;  V03 arg3         [V03    ] (  1,  1   )  struct (16) [fp+0x20]  do-not-enreg[XS] addr-exposed ld-addr-op <Point>
;  V04 arg4         [V04    ] (  1,  1   )  struct (16) [fp+0x10]  do-not-enreg[XS] addr-exposed ld-addr-op <Point>
;# V05 OutArgs      [V05    ] (  1,  1   )  struct ( 0) [sp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;
; Lcl frame size = 80
DEBUG: LOONGARCH64, frameType:1


G_M40687_IG01:              ;; offset=0x0000
  0xff7575a760  02FE8063  addi.d          sp, sp, -96
  0xff7575a764  29C00061  st.d            ra, sp, 0
  0xff7575a768  29C02076  st.d            fp, sp, 8
  0xff7575a76c  02C02076  addi.d          fp, sp, 8
  0xff7575a770  298152C4  st.w            a0, fp, 84 
  0xff7575a774  29C102C5  st.d            a1, fp, 64 
  0xff7575a778  29C122C6  st.d            a2, fp, 72 
  0xff7575a77c  29C0C2C7  st.d            a3, fp, 48 
  0xff7575a780  29C0E2C8  st.d            a4, fp, 56 
  0xff7575a784  29C082C9  st.d            a5, fp, 32 
  0xff7575a788  29C0A2CA  st.d            a6, fp, 40 
  0xff7575a78c  29C042CB  st.d            a7, fp, 16 
  0xff7575a790  28C1806C  ld.d            t0, sp, 96      // the old-style ABI will copy, but now the new stype not.
  0xff7575a794  29C062CC  st.d            t0, fp, 24      // the old-style ABI will copy, but now the new stype not. 
````

Reply:
I want to push a PR to fix this within the `CodeGen::genHomeRegisterParams`, is it needed ?

Reply:
```asm
  0xff7575a760  02FE8063  addi.d          sp, sp, -96
...
  0xff7575a790  28C1806C  ld.d            t0, sp, 96
```
The stack segment starts at +0. LA64 can allocate the address of `V04` to be `sp+88` instead (this PR should be doing that and I would assume the old logic also did that, unless other LA64/RV64 frame layout ends up overriding it). Then no stack segments need to be moved. This would make LA64 similar to arm32 and arm64.

Reply:
See https://github.com/dotnet/runtime/pull/101288#discussion_r1574689095 for an example of how it works on arm64. We only need to move the register segment because we allocate `V04` below the stack segment.

Reply:
> I want to push a PR to fix this within the `CodeGen::genHomeRegisterParams`, is it needed ?

This should not be fixed in `genHomeRegisterParams`. There are two approaches:
* Follow arm64 and arm32 and allocate the parameter below its stack segment, if that is possible. This is my preferred approach to make LA64/RV64 similar to arm32/arm64. It makes the stack segment copy unnecessary.
* Wait for #101288 to enable `genHomeSwiftStructParameters` for RV64/LA64. It will handle this problem in general.

Reply:
> What the `genHomeSwiftStructParameters` doing ？ I will check.

It is for `genHomeSplistStructParameters` ？

Reply:
If #101288 can fix this, I can wait.
Now our local runtime is fixed by workround within the `genHomeRegisterParams` by supporting split.

Reply:
> This makes the PSP's and monitor's offset is lager and make their offset computation complicated.
I think this is not the best.

It would be best to make this similar to arm32/arm64. It will avoid similar surprises in the future.
What is the complication here? LA64/RV64 already know how to handle parameters passed on stack only, so I would not expect it to be hard to extend this handling to work for the split parameters too. The correct offset is already assigned by `lvaAssignVirtualFrameOffsetsToArgs` (at least after #101340, but if I read the code right, also before).

Reply:
> > * Wait for [[RISC-V] Use common CodeGen::genHomeRegisterParams #101288](https://github.com/dotnet/runtime/pull/101288) to enable `genHomeSwiftStructParameters` for RV64/LA64. It will handle this problem in general.
> 
> What the `genHomeSwiftStructParameters` doing ？ I will check.

The `genHomeSwiftStructParameters` is complicated when solving the split case, we can resolve it immediately and no need to traval the whole args list againe.

Reply:
Iterating parameters is essentially free. There is no need to try to optimize this by adding special handling for LA64 or RV64 only. I would be much more worried about the unnecessary stack segment copy than enabling that code to run for LA64 or RV64.

Reply:
> > This makes the PSP's and monitor's offset is lager and make their offset computation complicated.
> > I think this is not the best.
> 
> It would be best to make this similar to arm32/arm64. It will avoid similar surprises in the future. What is the complication here? LA64/RV64 already know how to handle parameters passed on stack only, so I would not expect it to be hard to extend this handling to work for the split parameters too. The correct offset is already assigned by `lvaAssignVirtualFrameOffsetsToArgs` (at least after this PR, but if I read the code right, also before).

Just for split, your design and my design are both OK.
This is a tradeoff. I didn't see which is more better. just you like the arm32/64 and I like another.

Reply:
> Iterating parameters is essentially free. There is no need to try to optimize this by adding special handling for LA64 or RV64 only. I would be much more worried about the unnecessary stack segment copy than enabling that code to run for LA64 or RV64.

OK, I can accept each one which can process the split right.

Reply:
How toresolve it right now? 
within the #101288 by the `genHomeSwiftStructParameters` ?
or `The stack segments are always at the beginning and will be contiguous with the register segments. I thought LA64/RV64 was similar here.` ?

Reply:
I would prefer to have 1 design to keep in mind rather than 2 designs, but given that the Swift handling exists today I would be ok with enabling it for LA64/RV64 if you prefer this (cc @tomeksowi as well since you mostly have the change already done). It is general enough to handle arbitrary calling convention details, even if optimizations like in this case are possible.

FWIW, the other design is not usable for arm32 or arm64. If needs to work this way to make varargs work.

When you make LA64/RV64 different from other targets it makes it hard for us to make changes in common code. In this case I wasn't aware of this difference in `genFnPrologCalleeRegArgs` so I ended up breaking LA64 when merging #101224. I missed your comment around the split parameters; in the future feel free to mark the PR as NO-MERGE or draft to avoid this.

Reply:
> for arm32/arm64 we only need to store the register part. The stack segments are always at the beginning and will be contiguous with the register segments. I thought LA64/RV64 was similar here.

If we are planning to implement the split by this way, is it ok to merge the #100962 first which the FP  is near the top frame and will be easier for the offset.

And then I can push and discuss with RV64 how to amend the later codes.

Reply:
> I would prefer to have 1 design to keep in mind rather than 2 designs,

There is a delay to see your latest reply.
Yes, I agree with you.

Reply:
I just merged #100962 now after the RISCV64 testing.

Reply:
> FWIW, the other design is not usable for arm32 or arm64. If needs to work this way to make varargs work.

The varargs for LA64 is easy.

> When you make LA64/RV64 different from other targets.

I think you misunderstand my meanning.
And this is not my meanning.
As my previous reply which there is a delay for see your latest reply and you see my reply.
I think we have the same Viewpoint here.

Reply:
> I just merged #100962 now after the RISCV64 testing.

Thanks very much.

Reply:
> So I want to know your planning and I fellow your plan.

> > for arm32/arm64 we only need to store the register part. The stack segments are always at the beginning and will be contiguous with the register segments. I thought LA64/RV64 was similar here.
> 
> If we are planning to implement the split by this way, is it ok to merge the #100962 first which the FP is near the top frame and will be easier for the offset.
> 
> And then I can push and discuss with RV64 how to amend the later codes.

This is my point.
The previous different discusses are just for how to get your target which are best for all CPU-arches.

And I want to supply some points to be care, for example, the order of the float and integer args.
If needed the amending discusses of `Compiler::lvaAssignVirtualFrameOffsetsToLocals` for all CPU-arches.
Because if we can discuss more factors, it is more conceivable to design a better for all-CPU-arches.

Reply:
Yes, we can definitely discuss. If you are able to simplify functions like `lvaAssignVirtualFrameOffsetsToLocal` we would be happy to take it. However, if it comes with codegen differences or changed stack layouts, then we need some justifications for changes like this (e.g. perf measurements, codegen diffs with superpmi.py asmdiffs, etc.)

Reply:
> If we use this design, maybe it is better to do this by a new PR before https://github.com/dotnet/runtime/pull/101288, then the 101288 is more easier.

Let's see what @tomeksowi is working on in #101288 first. `genHomeSplitStructParameters` will be unnecessary for LA64/RV64 if you can use the arm32/arm64 approach.

Reply:
> > If we use this design, maybe it is better to do this by a new PR before #101288, then the 101288 is more easier.
> 
> Let's see what @tomeksowi is working on in #101288 first. `genHomeSplitStructParameters` will be unnecessary for LA64/RV64 if you can use the arm32/arm64 approach.

Yes, so we can discuss and get a uniform design for all CPU-arches.

Reply:
> Yes, we can definitely discuss. If you are able to simplify functions like `lvaAssignVirtualFrameOffsetsToLocal` we would be happy to take it. However, if it comes with codegen differences or changed stack layouts, then we need some justifications for changes like this (e.g. perf measurements, codegen diffs with superpmi.py asmdiffs, etc.)

This will be discussed by a new PR if needed.
And I agree with you that this is related many other modules and should be careful.

Reply:
> > > > for arm32/arm64 we only need to store the register part. The stack segments are always at the beginning and will be contiguous with the register segments. I thought LA64/RV64 was similar here.
> > > If we use this design, maybe it is better to do this by a new PR before #101288, then the 101288 is more easier.
> > 
> > Let's see what @tomeksowi is working on in #101288 first. `genHomeSplitStructParameters` will be unnecessary for LA64/RV64 if you can use the arm32/arm64 approach.
> 
> Yes, so we can discuss and get a uniform design for all CPU-arches. @tomeksowi what about your viewpoint?
> 
> If we accept this design and it is also convenient for adding the varargs for LA64/RV64.

@jakobbotsch 
If LA64/RV64 are similar with ARM32/ARM64, the saving order of the float and integer args should be considered with in this PR's `Compiler::lvaAssignVirtualFrameOffsetsToArgs`, is it ?

Then after this PR merged, we adjust the LA64/RV64's args position by a new PR, then the 101288 is more easier.

Reply:
Can you give an example?

This PR uses the new-style ABI information to figure out which parameter is split. Do you think it is not sufficient? Can you try this PR locally and show some examples where it's wrong?

Reply:
> Can you give an example?
> 
> This PR uses the new-style ABI information to figure out which parameter is split. Do you think it is not sufficient? Can you try this PR locally and show some examples where it's wrong?

Can I reply you tomorrow for my time is night now and also for I should confirm it further.

Reply:
Yes, please take your time. I will wait with merging this PR.

Reply:
> Yes, please take your time. I will wait with merging this PR.

Ok, thanks.
I just leave a quick reply first and this might be not right as I don't confirm your PR's code carefully.
for arglist, 
the half front args of arglist are using the integer register and split, then using the float registers, that is, the split case is in middle of the arglist and how we make sure the args's saving order to keep the split register's offset is on the highest stack and contiguous with the caller's stack?

Reply:
For varargs we handle arm64 by prespilling all registers: 
```scala
;  V00 VarArgHandle [V00    ] (  1,  1   )    long  ->  [fp+0x20]  do-not-enreg[V] ld-addr-op
;  V01 arg0         [V01    ] (  1,  1   )  struct (16) [fp+0x28]  do-not-enreg[XSA] multireg-arg addr-exposed <NativeVarargTest.TwoLongStruct>
;  V02 arg1         [V02    ] (  1,  1   )  struct (16) [fp+0x38]  do-not-enreg[XSA] multireg-arg addr-exposed <NativeVarargTest.TwoLongStruct>
;  V03 arg2         [V03    ] (  1,  1   )  struct (16) [fp+0x48]  do-not-enreg[XSA] multireg-arg addr-exposed <NativeVarargTest.TwoLongStruct>
;  V04 arg3         [V04    ] (  1,  1   )  struct (16) [fp+0x58]  do-not-enreg[XSA] multireg-arg addr-exposed <NativeVarargTest.TwoLongStruct>
;  V05 arg4         [V05    ] (  1,  1   )  struct (16) [fp+0x68]  do-not-enreg[XS] addr-exposed <NativeVarargTest.TwoLongStruct>
;# V12 OutArgs      [V12    ] (  1,  1   )  struct ( 0) [sp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;
; Lcl frame size = 16

G_M48227_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x60]!
            stp     x0, x1, [sp, #0x20]
            stp     x2, x3, [sp, #0x30]
            stp     x4, x5, [sp, #0x40]
            stp     x6, x7, [sp, #0x50]     ; always spilled in varargs, regardless of fixed arguments
            mov     fp, sp
            str     xzr, [fp, #0x18]    // [V11 loc0]
            str     x0, [fp, #0x20]     // [V00 VarArgHandle]
            str     x1, [fp, #0x28]     // [V01 arg0]
            str     x2, [fp, #0x30]     // [V01 arg0+0x08]
            str     x3, [fp, #0x38]     // [V02 arg1]      ; handling for fixed parameters; we could skip this in varargs methods because of prespills above
...
```

There is special logic in `lvaAssignVirtualFrameOffsetsToLocal` to assign the fixed parameters to the right location:
https://github.com/dotnet/runtime/blob/0a220d534b30495ff332b3bd65a04889e66b44f7/src/coreclr/jit/lclvars.cpp#L7122-L7131

Reply:
Yes, I agree with you.
Thanks!

There is a mistake for my using `arglist` which my meaning is not the `varargs`. Sorry for misguide you.

My meaning for `arglist` is common case liking:
The args of a function:
```c
struct  struct_a {int a; int b; int c; int d;};
struct point {double x; double y;};
int funcA(int num, struct_a a; struct_a b; point X1; point X2; point X3; point X4; struct_a c; struct_a d; struct_a e; point X5);   // there will be hit the split and the split is in the middle of the args and using both integer and float registers.
```

The half part args of function are using the integer register and split, then using the float registers, that is, the split case is in middle of the arglist and how we make sure the args's saving order to keep the split register's offset is on the highest stack and contiguous with the caller's stack?

Reply:
> What does the new style ABI information show in the JITDUMP? This PR assigns stack offsets for the args based on that.

(1) If you just want the disasm for LA64, I will add them later.
(2) What I mean is that, if LA64/RV64 save the args on the top of the frame by this way,
> for arm32/arm64 we only need to store the register part. The stack segments are always at the beginning and will be contiguous with the register segments. I thought LA64/RV64 was similar here.

When the split case is in middle of the arglist and both fload and integer registers are used, how we make sure the args's saving order to keep the split register's offset is on the highest stack and contiguous with the caller's stack?
I didn't find these code within the new style ABI to do that.

So I just feedback to you and how we do next step to suppport LA64/RV64's split args.

Reply:
> > What does the new style ABI information show in the JITDUMP? This PR assigns stack offsets for the args based on that.
> 
> (1) If you just want the disasm for LA64, I will add them later. (2) What I mean is that, if LA64/RV64 save the args on the top of the frame by this way to fix the split errors,
> 
> > for arm32/arm64 we only need to store the register part. The stack segments are always at the beginning and will be contiguous with the register segments. I thought LA64/RV64 was similar here.
> 
> When the split case is in middle of the arglist and both fload and integer registers are used, how we make sure the args's saving order to keep the split register's offset is on the highest stack and contiguous with the caller's stack? I didn't find these code within the new style ABI to do that.
> 
> So I just feedback to you and how we do next step to suppport LA64/RV64's split args.

This is just a feedback but not impede this PR merge.
We can discuss the LA64/RV64's split by a other PR.

Reply:
I think I'm not fully understanding the case, that's why I'm asking for the JITDUMP. For
```cpp
int funcA(int num, struct_a a; struct_a b; point X1; point X2; point X3; struct_a c; struct_a d; point X4; struct_a e; point X5); 
```
can you share me this part (like in https://github.com/dotnet/runtime/pull/101288#discussion_r1574689095):
```scala
Parameter V00 ABI info: [00..08) reg x0
Parameter V01 ABI info: 2 segments
  [0] [00..08) reg x1
  [1] [08..16) reg x2
Parameter V02 ABI info: 2 segments
  [0] [00..08) reg x3
  [1] [08..16) reg x4
Parameter V03 ABI info: 2 segments
  [0] [00..08) reg x5
  [1] [08..16) reg x6
Parameter V04 ABI info: 2 segments
  [0] [00..08) reg x7
  [1] [08..16) stack @ +00
Parameter V05 ABI info: [00..16) stack @ +08
Parameter V06 ABI info: [00..16) stack @ +24
Parameter V07 ABI info: [00..16) stack @ +40
Parameter V08 ABI info: [00..16) stack @ +56
Parameter V09 ABI info: [00..16) stack @ +72
Parameter V10 ABI info: [00..16) stack @ +88
```
? It will help me understand what case we are dealing with here.

Reply:
Ok. 
Should I dump this info based on this PR ?
I have to upload it later as I have to compile it first.

Reply:
You don't need it based on this PR. This PR will just assign the offsets to arguments based on the ABI information.

Reply:
Is this Ok ?  
I test it on my local runtime which I workround the split case.
```
Parameter V00 ABI info: [00..04) reg a0
Parameter V01 ABI info: 2 segments
  [0] [00..08) reg a1
  [1] [08..16) reg a2
Parameter V02 ABI info: 2 segments
  [0] [00..08) reg a3
  [1] [08..16) reg a4
Parameter V03 ABI info: 2 segments
  [0] [00..08) reg f0
  [1] [08..16) reg f1
Parameter V04 ABI info: 2 segments
  [0] [00..08) reg f2
  [1] [08..16) reg f3
Parameter V05 ABI info: 2 segments
  [0] [00..08) reg f4
  [1] [08..16) reg f5
Parameter V06 ABI info: 2 segments
  [0] [00..08) reg a5
  [1] [08..16) reg a6
Parameter V07 ABI info: 2 segments
  [0] [00..08) reg a7
  [1] [08..16) stack @ +00
Parameter V08 ABI info: 2 segments
  [0] [00..08) reg f6
  [1] [08..16) reg f7
Parameter V09 ABI info: [00..16) stack @ +08
Parameter V10 ABI info: [00..16) stack @ +24

lvaGrabTemp returning 12 (V12 tmp0) (a long lifetime temp) called for OutgoingArgSpace.

--------------------------------------------------------------------------------
G_M47983_IG01:              ;; offset=0x0000
  0xff75f1a860  02FD4063  addi.d          sp, sp, -176
  0xff75f1a864  2BC00078  fst.d           f24, sp, 0x0
  0xff75f1a868  29C02061  st.d            ra, sp, 8
  0xff75f1a86c  29C04076  st.d            fp, sp, 16
  0xff75f1a870  02C04076  addi.d          fp, sp, 16
  0xff75f1a874  298052C0  st.w            zero, fp, 20
  0xff75f1a878  298272C4  st.w            a0, fp, 156
  0xff75f1a87c  29C222C5  st.d            a1, fp, 136
  0xff75f1a880  29C242C6  st.d            a2, fp, 144
  0xff75f1a884  29C1E2C7  st.d            a3, fp, 120
  0xff75f1a888  29C202C8  st.d            a4, fp, 128
  0xff75f1a88c  2BC1A2C0  fst.d           f0, fp, 104
  0xff75f1a890  2BC1C2C1  fst.d           f1, fp, 112
  0xff75f1a894  2BC162C2  fst.d           f2, fp, 88
  0xff75f1a898  2BC182C3  fst.d           f3, fp, 96
  0xff75f1a89c  2BC122C4  fst.d           f4, fp, 72
  0xff75f1a8a0  2BC142C5  fst.d           f5, fp, 80
  0xff75f1a8a4  29C0E2C9  st.d            a5, fp, 56
  0xff75f1a8a8  29C102CA  st.d            a6, fp, 64
  0xff75f1a8ac  29C0A2CB  st.d            a7, fp, 40
  0xff75f1a8b0  28C2C06C  ld.d            t0, sp, 176
  0xff75f1a8b4  29C0C2CC  st.d            t0, fp, 48
  0xff75f1a8b8  2BC062C6  fst.d           f6, fp, 24
  0xff75f1a8bc  2BC082C7  fst.d           f7, fp, 32
````

Reply:
Yes it's ok.

This PR will assign in `lvaAssignVirtualFrameOffsetsToArgs`:
V07: -8
V09: +8
V10: +24

However, `lvaAssignVirtualFrameOffsetsToLocals` overwrites the assignment for `V07`. I think @tomeksowi is working on making it so that `V07` is not overwritten by `lvaAssignVirtualFrameOffsetsToLocals` in #101288.

Reply:
I'm a bit apprehensive about merging a PR which breaks everything with split struct args.

Reply:
Do you mean this PR? I don't think this PR will affect any behavior for either LA64/RISCV64. The only break right now came from #101224.

The original code computes the split stack parameters at similar offsets as this PR does for both LA64/RISCV64:
https://github.com/dotnet/runtime/blob/bc9fc5a774d96f95abe0ea5c90fac48b38ed2e67/src/coreclr/jit/lclvars.cpp#L6352-L6362
(In fact, this was one of the primary reasons I thought LA64/RV64 already worked this way for split parameters -- but the stack offset set there is never used.)

Reply:
It's possible I misunderstood what that code there is doing. Regardless, I don't think there is an issue with this PR since the offset this PR might be assigning to split parameters are getting overwritten regardless.

Reply:
> It's possible I misunderstood what that code there is doing.

I think I didn't express clearly to make you understand. 

> Regardless, I don't think there is an issue with this PR since the offset this PR might be assigning to split parameters are getting overwritten regardless.

Yes, what I feedback is not about this PR and I don't think LA64/RV64's splitting error is about this PR.
We just discussed here with you.
You can proceed this PR with your plan.